### PR TITLE
docs: More docs for UdpSocket

### LINKED ISSUE #2886: Concurrent sending over UDP appears to be broken on master
**Version**
master

**Platform**
```
❯ uname -ar
Linux leshowbox 5.8.10-arch1-1 #1 SMP PREEMPT Thu, 17 Sep 2020 18:01:06 +0000 x86_64 GNU/Linux
```

**Description**
While writing some new docs for `UdpSocket` I was trying to come up with an example using `Arc` for concurrent sending similarly to how `split` was used before. In doing so I found that the behavior is not the same between two bits of code that I think ought to behave the same.

This is code that behaves incorrectly:
```
use tokio::{net::UdpSocket, sync::mpsc};
use std::{io, net::SocketAddr, sync::Arc};

#[tokio::main]
async fn main() -> io::Result<()> {
    let sock = Arc::new(UdpSocket::bind("0.0.0.0:8080".parse::<SocketAddr>().unwrap()).await?);
    let (tx, mut rx) = mpsc::channel::<(Vec<u8>, SocketAddr)>(1_000);

    let udp_send = sock.clone();
    tokio::spawn(async move {
        while let Some((bytes, addr)) = rx.recv().await {
            let len = udp_send.send_to(&bytes, &addr).await.unwrap();
            println!("{:?} bytes sent", len);
        }
    });

    let mut buf = [0; 1024];
    loop {
        let (len, addr) = sock.recv_from(&mut buf).await?;
        println!("{:?} bytes received from {:?}", len, addr);
        tx.send((buf[..len].to_vec(), addr)).await.unwrap();
    }
}
```

I expected to see this happen: it echoes responses

Instead, this happened: it hangs forever

---

To be clear, the original code with `split` was:
```

#[tokio::main]
async fn main() -> io::Result<()> {
    let sock = UdpSocket::bind("0.0.0.0:8080".parse::<SocketAddr>().unwrap()).await?;

    let (mut udp_recv, mut udp_send) = sock.split();
    let (mut tx, mut rx) = mpsc::channel::<(Vec<u8>, SocketAddr)>(1_000);

    tokio::spawn(async move {
        while let Some((bytes, addr)) = rx.recv().await {
            let len = udp_send.send_to(&bytes, &addr).await.unwrap();
            println!("{:?} bytes sent", len);
        }
    });

    let mut buf = [0; 1024];
    loop {
        let (len, addr) = udp_recv.recv_from(&mut buf).await?;
        println!("{:?} bytes received from {:?}", len, addr);
        tx.send((buf[..len].to_vec(), addr)).await;
    }
}
```

---

A test can be added which shows this behavior also, assuming a fix needs to be made I think this should will ensure it works:
```
#[tokio::test]
async fn split_chan() -> std::io::Result<()> {
    // setup UdpSocket that will echo all sent items
    let socket = UdpSocket::bind("127.0.0.1:0").await?;
    let addr = socket.local_addr().unwrap();
    let s = Arc::new(socket);
    let r = s.clone();

    let (tx, mut rx) = mpsc::channel::<(Vec<u8>, SocketAddr)>(1_000);
    tokio::spawn(async move {
        while let Some((bytes, addr)) = rx.recv().await {
            s.send_to(&bytes, &addr).await.unwrap();
        }
    });

    tokio::spawn(async move {
        let mut buf = [0u8; 32];
        loop {
            let (len, addr) = r.recv_from(&mut buf).await.unwrap();
            tx.send((buf[..len].to_vec(), addr)).await.unwrap();
        }
    });

    // test that we can send a value and get back some response
    let sender = UdpSocket::bind("127.0.0.1:0").await?;
    sender.send_to(MSG, addr).await?;
    let mut recv_buf = [0u8; 32];
    let (len, _) = sender.recv_from(&mut recv_buf).await?;
    assert_eq!(&recv_buf[..len], MSG);
    Ok(())
}
```

### PR DESCRIPTION
**Pull request by leshow**
# Motivation

I always found the udp docs a little lackluster, after using them extensively for some time I thought it might be nice to flesh it out a bit. I asked in the tokio-docs discord and others seemed open to it, so here they are! Would love to hear your thoughts

## Solution

A few small code examples similar to the way the tcp section is set up, with an explanation of the different ways to use `UdpSocket`.

~~I had a full larger example with `split` but found that it's been removed for `UdpSocket` in master, I hope that's temporary, if not is there anything I can do to get similar functionality back? i.e. being able to `send_to`/`recv_from` concurrently~~

~~In writing an example using the new way to share `UdpSocket` concurrently (post-`split`) I think I found a bug~~ this bug was fixed #2886

### CODE REVIEW

**Comment by leshow** [comment]
Does anyone else have issues with `futures_util` when running cargo test locally? I'm running from the `tokio` sub crate in the workspace:

```
~/dev/rust/tokio/tokio master
❯ cargo test --features full
   Compiling autocfg v0.1.7
   Compiling syn v1.0.41
   Compiling rand_core v0.4.2
   Compiling getrandom v0.1.13
   Compiling ppv-lite86 v0.2.6
   Compiling pin-project-internal v0.4.23
   Compiling log v0.4.8
   Compiling proc-macro-nested v0.1.3
   Compiling once_cell v1.4.1
   Compiling futures-sink v0.3.5
   Compiling futures-io v0.3.5
   Compiling remove_dir_all v0.5.2
   Compiling bitflags v1.2.1
   Compiling pin-utils v0.1.0
   Compiling byteorder v1.3.2
   Compiling bit-vec v0.5.1
   Compiling quick-error v1.2.2
   Compiling regex-syntax v0.6.18
   Compiling futures-channel v0.3.5
   Compiling wait-timeout v0.2.0
   Compiling futures-task v0.3.5
   Compiling rand_core v0.3.1
   Compiling rand_jitter v0.1.4
   Compiling rand_os v0.1.3
   Compiling bit-set v0.5.1
   Compiling rand_xorshift v0.1.1
   Compiling rand_hc v0.1.0
   Compiling rand_isaac v0.1.1
   Compiling rand_core v0.5.1
   Compiling mio v0.6.21
   Compiling rand_pcg v0.1.2
   Compiling rand_chacha v0.1.1
   Compiling rand v0.6.5
   Compiling num-traits v0.2.10
   Compiling c2-chacha v0.2.3
   Compiling mio-uds v0.6.7
   Compiling rand_chacha v0.2.1
   Compiling rand v0.7.2
   Compiling tempfile v3.1.0
   Compiling rusty-fork v0.2.2
   Compiling proptest v0.9.4
   Compiling proc-macro-hack v0.5.11
   Compiling tokio-macros v0.3.0 (/home/leshow/dev/rust/tokio/tokio-macros)
   Compiling tokio v0.3.0 (/home/leshow/dev/rust/tokio/tokio)
   Compiling futures-macro v0.3.5
   Compiling pin-project v0.4.23
   Compiling futures-util v0.3.5
   Compiling tokio-test v0.3.0 (/home/leshow/dev/rust/tokio/tokio-test)
   Compiling futures-executor v0.3.5
   Compiling futures v0.3.1
error[E0433]: failed to resolve: could not find `document_join_macro` in `futures_util`
   --> /home/leshow/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.3.1/src/lib.rs:504:15
    |
504 | futures_util::document_join_macro! {
    |               ^^^^^^^^^^^^^^^^^^^ could not find `document_join_macro` in `futures_util`

error[E0433]: failed to resolve: could not find `document_select_macro` in `futures_util`
   --> /home/leshow/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.3.1/src/lib.rs:528:15
    |
528 | futures_util::document_select_macro! {
    |               ^^^^^^^^^^^^^^^^^^^^^ could not find `document_select_macro` in `futures_util`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0433`.
error: could not compile `futures`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by taiki-e** [comment]
> Does anyone else have issues with `futures_util` when running cargo test locally? I'm running from the `tokio` sub crate in the workspace:

it could be fixed by run `cargo update`.

**Comment by leshow** [comment]
Ah great, that did the trick. I realized that the api has been changed so `UdpSocket` is no longer `mut`. I suppose that's why `split` is gone?

I'd like to include another example of putting `UdpSocket` in an `Arc` and using it concurrently if that's alright. Since it was not obvious to me that it was even possible to do that, having been used to the `split` API before.

**Review by carllerche — APPROVED** [review]
thanks :+1:

File: tokio/src/net/udp/socket.rs [hunk 1]
```diff
@@ -183,6 +380,22 @@ impl UdpSocket {
...
+    ///   # use std::{io, net::SocketAddr};
+    ///
+    /// #   #[tokio::main]
+    /// #   async fn main() -> io::Result<()> {
+    ///        let sock = UdpSocket::bind("0.0.0.0:8080".parse::<SocketAddr>().unwrap()).await?;
+    ///        let mut buf = [0u8; 32];
+    ///        let (len, addr) = sock.recv_from(&mut buf).await?;
+    ///        println!("received {:?} bytes from {:?}", len, addr);
+    ///  #      Ok(())
+    ///  #  }
```
Comment (taiki-e) [inline tokio/src/net/udp/socket.rs:397]:
```suggestion
    /// use tokio::net::UdpSocket;
    /// # use std::{io, net::SocketAddr};
    ///
    /// # #[tokio::main]
    /// # async fn main() -> io::Result<()> {
    /// let sock = UdpSocket::bind("0.0.0.0:8080".parse::<SocketAddr>().unwrap()).await?;
    /// let mut buf = [0u8; 32];
    /// let (len, addr) = sock.recv_from(&mut buf).await?;
    /// println!("received {:?} bytes from {:?}", len, addr);
    /// # Ok(())
    /// # }
```

**Review by taiki-e — CHANGES_REQUESTED** [review]

**Review by taiki-e — APPROVED** [review]

**Pull request merged**